### ISSUE TITLE
vinyl: avoid linear scans for duplicate elimination during dump

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#2499]
### Duplicate elimination during dump

The new write iterator, when dumps an in-memory tree (or trees) to disk uses a linear scan over all tuples versions to position to the next key. For very hot keys, this may use up quite a bit of CPU unnecessarily. Some hot keys can take millions of updates before a dump. Avoid linear scans for duplicate elimination of such keys.

The following alternatives are possible: 

- re-introduce some kind of next_key() to the write iterator source interface (aka stream)
- eliminate duplicates from the trees at update time. We can implement a technique
similar to upsert squashing to purge old keys from the tree right away if they are not used in any of the read views.

We need to weigh pro & cons of each of the two approaches (or perhaps find the third one).

The advantage of the first approach is that the cost of duplicate elimintation is logN lookup in the tree **in the dump thread**.

With the second approach, we lack a way to cut out a piece of bps tree cheaply: we still need to delete elements one by one, i.e. linearly, and this time **in tx thread**. This is a serious drawback.

We perhaps could implement an API in bps which would allow us to cut out an entire block at a constant cost. Then we would simply trigger the cut out when we have an entire block full of duplicate tuples. 
The possible advantage of the second approach is that in-memory tree stays more compact for lookups. This doesn't seem to be a big deal, since in BPS tree the number of lookups is only log(N).

### Duplicate elimination during compaction

The above problem doesn't seem to have a serious performance impact on compaction, since during compaction we eliminate duplicate for stale read views only. We don't expect to have more than a hundred or perhaps few hundreds of read views, and most likely these read views will  not require keeping even a handful of versions, even for hottest tuples.

**Comment by alyapunov** [comment]
>re-introduce some kind of next_key() to the write iterator source interface (aka stream):

Having a mem with N tuples means we have already spent O(N log(N)) time in TX thread building it. After that skipping even N tuples in worker thread doesn't matter since we dump a mem only once. Even a complete optimization will not be seen in overall CPU usage. It also must be clear that skipping a significant number of tuples during a dump is already the best possible case for a dumper.

 Nevertheless it can be optimized but I suggest lowest priority.

>eliminate duplicates from the trees at update time

Upsert squash is important because user can select same key any amount of times. Unlike upserts, tons of replaces of the same key cannot affect point queries' performance. They can affect range scans, and read iterators must handle the case properly (now they don't). Removing one useless statement from mem's tree
 - will cost O(log(N)) in TX instead of  O(1) in dump worker.
 - will not free memory because of lsregion.
 - can improve logarithmic complexity of a tree in case of domination of hot data over cold data (Memtx must be used in this case).

So I guess removing useless is useless.

See #2500

**Comment by kostja** [comment]
Won't fix.
